### PR TITLE
Gradient function.

### PR DESCRIPTION
**Pull request by odashi**
This change adds `node::grad()`, which calculates gradient (partial derivative) values `\partial y / \partial x` given `y` and `x`.
The function extends the existing `Graph` to explain subgraphs for the gradient calculation and returns new `Node`s representing the gradients.
By using the returned `Node` recursively, `grad()` can be used to explain higher-order gradients.

To implement this functionality, this change also fixes an ownership problem around `Graph` and `Operator`.
Specifically, the `Operator` trait now provides `get_gradient_fn()` instead of `gradient()`, which returns a `Gradient` trait object to represent the operator's gradient function, while the `gradient()` requires to manipulate `Graph` which results in runtime violation of borrowing.

Blocking change: #16

### CODE REVIEW

File: src/graph.rs [hunk 1]
```diff
@@ -22,11 +24,23 @@ impl<'hw> ArrayPlaceholder<'hw> {
...
 
+    /// Obtains the `Hardware` of this placeholder.
+    ///
+    /// # Returns
+    ///
+    /// A reference to the `Hardware` object.
+    pub(crate) fn hardware(&self) -> &'hw RefCell<dyn Hardware> {
+        match self {
+            Self::Unassigned(_, hardware) => hardware,
+            Self::Assigned(ref array) => array.hardware(),
```
Comment (vbkaisetsu) [inline src/graph.rs]: Is `ref` necessary? It is used in many places, but I think most of them are unnecessary.
```suggestion
            Self::Assigned(array) => array.hardware(),
```

Comment (odashi) [reply src/graph.rs]: Yes in some cases the destructured thing is already a reference, so it is unnecessary.

I guess there is some ambiguity of the language itself in the context of destructuring, so the compiler couldn't teach me about this redundancy: 🤔 

Untitled

Comment (odashi) [reply src/graph.rs]: fixed fc1fd0a

Comment (odashi) [reply src/graph.rs]: Oh the ambiguity is explicitly introduced,
https://github.com/rust-lang/rfcs/blob/master/text/2005-match-ergonomics.md

**Comment by tetsuok** [comment]
@odashi would you resolve merge conflicts?

**Comment by odashi** [comment]
@tetsuok Sorry I didn't have time to do it, will do soon.

**Comment by odashi** [comment]
@tetsuok @vbkaisetsu please review this pr again.

**Review by vbkaisetsu — APPROVED** [review]

**Review by tetsuok — APPROVED** [review]

**Pull request merged**